src/main/java/hudson/remoting/PingThread.java

```diff
@@ -103,13 +103,13 @@ public void run() {
     private void ping() throws IOException, InterruptedException {
         Future<?> f = channel.callAsync(new Ping());
         long start = System.currentTimeMillis();
-        long end = start +timeout;
 
-        long remaining;
+        long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout);
+        long remaining = end - System.nanoTime();
+
         do {
-            remaining = end-System.currentTimeMillis();
             try {
-                f.get(Math.max(0,remaining),MILLISECONDS);
+                f.get(Math.max(TimeUnit.MILLISECONDS.toNanos(10),remaining),TimeUnit.NANOSECONDS);
```

Comment:
@stephenc if the channel is closed during this call then f.get(...) should throw an exception if I follow the code correctly.
So this should return as soon as
1) the result is available
or
2) the channel is closed
or
3) The timeout happens
or
4) the JVM does a spurious wakeup...